src/sage/rings/polynomial/polynomial_element.pyx

```diff
@@ -2560,22 +2583,69 @@ cdef class Polynomial(CommutativePolynomial):
                 return rs[0]
             raise ValueError(f"polynomial {self} has no roots")
 
+        # Ensure that a provided ring is appropriate for the function
+        if ring is not None:
+            # If the new ring is not a finite field, attempt to coerce the polynomial
+            # and call the function to use naive factoring
+            if ring not in FiniteFields():
+                try:
+                    f = self.change_ring(ring)
+                except ValueError:
+                    raise(f"cannot coerce polynomial {self} to the new ring: {ring}")
+                return f.any_root()
+
+            # If the ring is a finite field, ensure it's an extension of the base ring
+            if self.base_ring().characteristic() != ring.characteristic():
+                raise ValueError("ring must have the same characteristic as the base ring")
+            if not self.base_ring().degree().divides(ring.degree()):
```

Comment:
This is not sufficient to ensure there's a coercion. Try `if *.coerce_map_from(*) is None`.

Reply:
Done